shared/src/balancer/event_handler.rs

```diff
@@ -77,24 +83,43 @@ trait PoolDataFetching: Send + Sync {
 impl PoolDataFetching for Web3 {
     /// Could result in ethcontract::{NodeError, MethodError or ContractError}
     async fn get_pool_data(&self, pool_address: H160) -> Result<WeightedPoolData> {
+        let mut batch = CallBatch::new(self.transport());
         let pool_contract = BalancerV2WeightedPool::at(self, pool_address);
         // Need vault and pool_id before we can fetch tokens.
         let vault = BalancerV2Vault::deployed(&self).await?;
         let pool_id = H256::from(pool_contract.methods().get_pool_id().call().await?.0);
-        let tokens = vault
+
+        // token_data and weight calls can be batched
+        let token_data = vault
             .methods()
             .get_pool_tokens(Bytes(pool_id.0))
-            .call()
-            .await?
-            .0;
+            .batch_call(&mut batch);
+        let normalized_weights = pool_contract
+            .methods()
+            .get_normalized_weights()
+            .batch_call(&mut batch);
+        batch.execute_all(MAX_BATCH_SIZE).await;
+
+        let tokens = token_data.await?.0;
+        // This is already a batch call for a list of token decimals.
+        let token_decimals = TokenInfoFetcher { web3: self.clone() }
+            .get_token_infos(&tokens)
+            .await;
+        let scaling_factors = token_decimals
+            .iter()
+            .map(|(_, info)| {
+                // Note that balancer does not support tokens with more than 18 decimals
+                // https://github.com/balancer-labs/balancer-v2-monorepo/blob/ce70f7663e0ac94b25ed60cb86faaa8199fd9e13/pkg/pool-utils/contracts/BasePool.sol#L497-L508
+                18 - info
+                    .decimals
+                    .expect("all token decimals must be available build pool data")
```

Comment:
Maybe it's better to return an error here, otherwise you could create a token with customizable decimals, create a Balancer pool with the decimals set to 18, then increase the decimals and make the services crash.

Reply:
I thought that the expect message was the error... Perhaps I have misunderstood. I suppose expect panics when it fails.

Reply:
`expect` panics with that error message, but since this is something that could be used to make the services crash we could make `get_pool_data` return an error in this case instead.